x/gamm/pool-models/stableswap/amm.go

```diff
@@ -352,11 +356,51 @@ func (pa *Pool) calcInAmtGivenOut(tokenOut sdk.Coin, tokenInDenom string, swapFe
 	if err != nil {
 		return sdk.Dec{}, err
 	}
-	tokenInSupply := reserves[0].ToDec()
-	tokenOutSupply := reserves[1].ToDec()
+	tokenInSupply, tokenOutSupply := reserves[0], reserves[1]
 	// We are solving for the amount of token in, cfmm(x,y) = cfmm(x + x_in, y - y_out)
 	// x = tokenInSupply, y = tokenOutSupply, yIn = -tokenOutAmount
 	cfmmIn := solveCfmm(tokenInSupply, tokenOutSupply, tokenOut.Amount.ToDec().Neg())
 	inAmt := pa.getDescaledPoolAmt(tokenInDenom, cfmmIn.NegMut())
 	return inAmt, nil
 }
+
+func (pa *Pool) calcSingleAssetJoinShares(tokenIn sdk.Coin, swapFee sdk.Dec) (sdk.Int, error) {
+	poolWithAddedLiquidityAndShares := func(newLiquidity sdk.Coin, newShares sdk.Int) types.PoolI {
+		paCopy := pa.Copy()
+		paCopy.updatePoolForJoin(sdk.NewCoins(tokenIn), newShares)
+		return &paCopy
+	}
+	// TODO: Correctly handle swap fee
+	return cfmm_common.BinarySearchSingleAssetJoin(pa, tokenIn, poolWithAddedLiquidityAndShares)
+}
+
+// We can mutate pa here
+// TODO: some day switch this to a COW wrapped pa, for better perf
+func (pa *Pool) joinPoolSharesInternal(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error) {
+	if len(tokensIn) == 1 {
+		numShares, err = pa.calcSingleAssetJoinShares(tokensIn[0], swapFee)
+		newLiquidity = tokensIn
+		return numShares, newLiquidity, err
+	} else if len(tokensIn) != pa.NumAssets() {
+		return sdk.ZeroInt(), sdk.NewCoins(), errors.New(
+			"stableswap pool only supports LP'ing with one asset, or all assets in pool")
+	}
```

Comment:
nit: `else if` adds verbosity for input checking
```suggestion
	}
	if len(tokensIn) != pa.NumAssets() {
		return sdk.ZeroInt(), sdk.NewCoins(), errors.New(
			"stableswap pool only supports LP'ing with one asset, or all assets in pool")
	}
```

Reply:
I don't follow? Why would you not prefer else if?

Reply:
My general view on this sort of code linting is either should have a readability reason, or a linter enforce it

Reply:
My reasoning for this comment was the following:

In the case of if/else, we only execute A if condA is true
```
if condA {
   // Do A
} else if condB {
   // Do B
}
```
On the other hand,
```
if condA {
   // Do A
}

if condB {
   // Do B
}
```

It is possible to do both A and B independent of the value of `condA`.

Functionally, it does not make a difference in this example since both branches return. However, when reading the code, there is a small readability overhead for validating this subtle difference between `if/else` and `if`. Especially, since Go frequently uses the latter pattern for error handling.

I don't have a strong opinion on this so feel free to keep what you think is best :+1: